# Assignment 1

In [72]:
import numpy as np
import pandas as pd
import sklearn as sl
import matplotlib as mp
import scipy as cp
import seaborn as sb
from matplotlib import pyplot as plt
import statsmodels.api as sm
from sklearn.datasets import load_iris
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import f1_score
from sklearn import linear_model as lm
from sklearn.model_selection import GridSearchCV
from sklearn.metrics import make_scorer

## Part A

1.	Download the “dataset_lm.csv” file from Canvas and upload it to Jupyter Notebook. 

In [47]:
df = pd.read_csv(r"C:\Users\johns\OneDrive\Desktop\MBAN Semester 2\Predictive Analytics\Assignment 1\dataset_lm.csv")

In [48]:
df.head()

Dependent Var  Explanatory Var #1  Explanatory Var #2  Explanatory Var #3  \
0      56.293458           13.698667           50.639873                   0   
1      58.473431            2.714725           65.845845                   1   
2      94.195330           11.618072           65.072497                   0   
3      29.074583            0.818623           45.408996                   1   
4      86.035569            9.077544           73.548021                   0   

   Explanatory Var #4  Explanatory Var #5  Explanatory Var #6  \
0          -18.568035           45.121911           11.412501   
1          -25.105932           47.190213           10.080280   
2           -7.897464           52.163036           11.057301   
3          -18.316132           54.356714            5.029029   
4          -19.204165           47.186807           12.128134   

   Explanatory Var #7  Explanatory Var #8  Explanatory Var #9  \
0           56.410757                   2          -12.281132   
1           65.383107                   3          -36.763585   
2           82.812717                   0          -15.733547   
3           48.812471                   1          -12.825591   
4           62.520911                   2          -13.804860   

   Explanatory Var #10  Explanatory Var #11  Explanatory Var #12  \
0            38.996909            -3.010548            49.195073   
1            51.654939             4.991111            45.591729   
2            48.913837            -2.457696            56.608806   
3            45.851732            14.974177            47.362594   
4            47.765904             9.593982            53.700562   

   Explanatory Var #13  Explanatory Var #14  Explanatory Var #15  
0                    0           -21.153143            46.919314  
1                    0            -6.474403            53.383508  
2                    0           -27.903299            48.515026  
3                    1           -10.064411            55.266254  
4                    0           -17.546302            48.150543

In [49]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 422 entries, 0 to 421
Data columns (total 16 columns):
 #   Column               Non-Null Count  Dtype  
---  ------               --------------  -----  
 0   Dependent Var        422 non-null    float64
 1   Explanatory Var #1   422 non-null    float64
 2   Explanatory Var #2   422 non-null    float64
 3   Explanatory Var #3   422 non-null    int64  
 4   Explanatory Var #4   422 non-null    float64
 5   Explanatory Var #5   422 non-null    float64
 6   Explanatory Var #6   422 non-null    float64
 7   Explanatory Var #7   422 non-null    float64
 8   Explanatory Var #8   422 non-null    int64  
 9   Explanatory Var #9   422 non-null    float64
 10  Explanatory Var #10  422 non-null    float64
 11  Explanatory Var #11  422 non-null    float64
 12  Explanatory Var #12  422 non-null    float64
 13  Explanatory Var #13  422 non-null    int64  
 14  Explanatory Var #14  422 non-null    float64
 15  Explanatory Var #15  422 non-null    flo

2.	Run the OLS model by using the dependent and explanatory variables in the dataset. 

In [50]:
y = df['Dependent Var']
X = df.drop('Dependent Var', axis = 1)

In [51]:
model_reg = sm.OLS(y,X).fit()

3.	Show your summary table in Python and interpret your results in the summary report.

In [52]:
model_reg.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                                 OLS Regression Results                                
=======================================================================================
Dep. Variable:          Dependent Var   R-squared (uncentered):                   0.999
Model:                            OLS   Adj. R-squared (uncentered):              0.999
Method:                 Least Squares   F-statistic:                          4.509e+04
Date:                Mon, 16 Oct 2023   Prob (F-statistic):                        0.00
Time:                        19:36:24   Log-Likelihood:                         -841.76
No. Observations:                 422   AIC:                                      1714.
Df Residuals:                     407   BIC:                                      1774.
Df Model:                          15                                                  
Covariance Type:            nonrobust                                                  
=======================================================================================
                          coef    std err          t      P>|t|      [0.025      0.975]
---------------------------------------------------------------------------------------
Explanatory Var #1      1.3078      0.013    102.159      0.000       1.283       1.333
Explanatory Var #2      1.7667      0.009    203.101      0.000       1.750       1.784
Explanatory Var #3      6.6083      0.177     37.232      0.000       6.259       6.957
Explanatory Var #4      2.0725      0.011    189.896      0.000       2.051       2.094
Explanatory Var #5     -0.7776      0.011    -68.372      0.000      -0.800      -0.755
Explanatory Var #6      0.0124      0.008      1.595      0.111      -0.003       0.028
Explanatory Var #7      0.0286      0.006      4.890      0.000       0.017       0.040
Explanatory Var #8      0.3510      0.080      4.398      0.000       0.194       0.508
Explanatory Var #9     -0.0293      0.010     -2.838      0.005      -0.050      -0.009
Explanatory Var #10     0.1292      0.011     11.409      0.000       0.107       0.152
Explanatory Var #11     0.0145      0.013      1.152      0.250      -0.010       0.039
Explanatory Var #12     0.0752      0.009      8.698      0.000       0.058       0.092
Explanatory Var #13     0.7414      0.177      4.198      0.000       0.394       1.089
Explanatory Var #14    -0.0218      0.011     -1.974      0.049      -0.043   -8.61e-05
Explanatory Var #15     0.1212      0.012     10.324      0.000       0.098       0.144
==============================================================================
Omnibus:                        0.087   Durbin-Watson:                   1.875
Prob(Omnibus):                  0.957   Jarque-Bera (JB):                0.188
Skew:                          -0.004   Prob(JB):                        0.910
Kurtosis:                       2.897   Cond. No.                         287.
==============================================================================

Notes:
[1] R² is computed without centering (uncentered) since the model does not contain a constant.
[2] Standard Errors assume that the covariance matrix of the errors is correctly specified.
"""

## Part B

1.	Use error values from the OLS model to calculate their standard deviation and autocorrelation values for the first three lags. 

In [53]:
pred_vals = model_reg.predict(X)
pred_vals

0      51.467745
1      58.671234
2      93.220317
3      27.220630
4      84.924624
         ...    
417    64.536548
418    26.764015
419    55.498955
420    34.601013
421    49.555741
Length: 422, dtype: float64

In [54]:
st_dev = np.std(pred_vals)
st_dev

25.84534616492947

In [55]:
lags = 3
autocorrelation = [np.corrcoef(df[i:], df[:-i])[0, 1] for i in range(1, lags + 1)]
autocorrelation

[0.9338889427200325, 0.8568714054990086, 0.8841512562268686]

2.	Then, run the GLS model accordingly. 

In [56]:
gls_model = sm.GLS(y, X)

In [59]:
order = 1
cov_structure = sm.cov_struct.Autoregressive(order)
gls_model.cov_struct = cov_structure

In [60]:
gls_results = gls_model.fit()

3.	Show your summary table in Python and interpret your results in the summary report.

In [61]:
print(gls_results.summary())

                                 GLS Regression Results                                
Dep. Variable:          Dependent Var   R-squared (uncentered):                   0.999
Model:                            GLS   Adj. R-squared (uncentered):              0.999
Method:                 Least Squares   F-statistic:                          4.509e+04
Date:                Mon, 16 Oct 2023   Prob (F-statistic):                        0.00
Time:                        19:37:47   Log-Likelihood:                         -841.76
No. Observations:                 422   AIC:                                      1714.
Df Residuals:                     407   BIC:                                      1774.
Df Model:                          15                                                  
Covariance Type:            nonrobust                                                  
                          coef    std err          t      P>|t|      [0.025      0.975]
--------------------------------

## Part C

1.	Split the dataset into two as the training and test sets (test size = 0.5). 

In [63]:
from sklearn.model_selection import train_test_split

In [64]:
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.5, random_state=42)

2.	Run the Lasso model with alpha=1 and estimate the coefficients using the training set. 

In [67]:
model_lasso = lm.Lasso(alpha=1).fit(X_train,y_train)
model_lasso.coef_

array([ 1.26972628,  1.68394638,  2.02626245,  2.08756512, -0.91746375,
       -0.        ,  0.        , -0.        ,  0.        , -0.        ,
        0.01314162,  0.        , -0.        ,  0.        , -0.03617731])

3.	Then, calculate the mean absolute percentage error using the test set. 

In [68]:
predictions = model_lasso.predict(X_test)

In [69]:
absolute_percentage_errors = abs((y_test - predictions) / y_test) * 100

In [70]:
mape = absolute_percentage_errors.mean()    

In [71]:
print("Mean Absolute Percentage Error (MAPE):", mape)

Mean Absolute Percentage Error (MAPE): 4.432190198291583


4.	Find an approximate value for alpha that minimizes the mean absolute percentage error.

In [75]:
model_lasso = lm.Lasso(alpha=0.01).fit(X_train,y_train)
model_lasso.coef_

predictions = model_lasso.predict(X_test)
absolute_percentage_errors = abs((y_test - predictions) / y_test) * 100
mape = absolute_percentage_errors.mean()    
print("Mean Absolute Percentage Error (MAPE) for Alpha 0.01:", mape)

Mean Absolute Percentage Error (MAPE) for Alpha 0.01: 0.04429881163345532


In [76]:
model_lasso = lm.Lasso(alpha=0.1).fit(X_train,y_train)
model_lasso.coef_

predictions = model_lasso.predict(X_test)
absolute_percentage_errors = abs((y_test - predictions) / y_test) * 100
mape = absolute_percentage_errors.mean()    
print("Mean Absolute Percentage Error (MAPE) for Alpha 0.1:", mape)

Mean Absolute Percentage Error (MAPE) for Alpha 0.1: 0.44323186549579874


In [77]:
model_lasso = lm.Lasso(alpha=10).fit(X_train,y_train)
model_lasso.coef_

predictions = model_lasso.predict(X_test)
absolute_percentage_errors = abs((y_test - predictions) / y_test) * 100
mape = absolute_percentage_errors.mean()    
print("Mean Absolute Percentage Error (MAPE) for Alpha 10:", mape)

Mean Absolute Percentage Error (MAPE) for Alpha 10: 8.587094351590437


In [78]:
model_lasso = lm.Lasso(alpha=100).fit(X_train,y_train)
model_lasso.coef_

predictions = model_lasso.predict(X_test)
absolute_percentage_errors = abs((y_test - predictions) / y_test) * 100
mape = absolute_percentage_errors.mean()    
print("Mean Absolute Percentage Error (MAPE) for Alpha 100:", mape)

Mean Absolute Percentage Error (MAPE) for Alpha 100: 44.287702093738424


Mean Absolute Percentage Error (MAPE) for Alpha 0.01: 0.04429881163345532

Mean Absolute Percentage Error (MAPE) for Alpha 0.1: 0.44323186549579874

Mean Absolute Percentage Error (MAPE) for Alpha 1: 4.432190198291583

Mean Absolute Percentage Error (MAPE) for Alpha 10: 8.587094351590437

Mean Absolute Percentage Error (MAPE) for Alpha 100: 44.287702093738424

## Part D

1.	Use the demand data given in the table and develop an appropriate forecasting model (i.e., the tailored regularization discussed in the class—see your slides for more info) that exploits the available information given in the table as much as possible.